### PR TITLE
Emulator squash

### PR DESCRIPTION
**Pull request by nbren12**
Squashing small cloud values to zero was an idea to prevent negative cloud water from forming without biasing the model too much. This is a set of experiments and some related generally useful change (it's a bit of a grab bag).

Empirically it did not help much, but it might be useful later on, so we may as well merge the option (disabled by default).

Added public API:
- projects/microphysics/scripts/query-wandb.py. a utility for querying top level prognostic run metrics from wandb. Hopefully useful for other sets of experiments labelled by single tag.
- `prognostic_run_diags piggy --summary-filter`. allows subselecting summary metric computations for faster computation.
- New options: [ModelConfig.{cloud_squash,gscond_cloud_conservative}](https://github.com/ai2cm/fv3net/pull/1771/commits/4ea47b46be1a49feb19a4bec483d2c8b99a44d24)

Significant internal changes:
- Fixed fv3net.diagnostics.prognostic_run.logs. parse_duration to also work with json-formatted FV3 logs
- Pin ubuntu version to specific digest. Hopefully this will improve the prognostic run build time.

Requirement changes:
- add typer as input
- add `projects/microphysics/requirements.in` for tooling dependencies of the microphysics project.

### CODE REVIEW

File: projects/microphysics/scripts/query-wandb.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+@app.command()
+def prognostic_runs(experiment: str, filter_tags: str = typer.Option("", "-f")):
+    """Show the top level metrics for prognostic runs tagged by `experiment`
+
+    Examples:
+
+    Rerun the piggy-backed diagnostics for all runs in an experiment::
+
+        top.py prognostic-runs experiment/squash -f bug \
```
Comment (frodre) [inline projects/microphysics/scripts/query-wandb.py:30]: `top.py` -> `query-wandb.py`

File: projects/microphysics/scripts/query-wandb.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+
+@app.command()
+def runs(experiment: str, filter_tags: str = typer.Option("", "-f")):
+    """
+    Examples:
+
+    Filter all runs with "bug" in tags::
+
+        ./top.py runs experiment/squash -f bug | jq -sr '[.[].group] | unique | .[]'
```
Comment (frodre) [inline projects/microphysics/scripts/query-wandb.py:83]: `top.py` -> `query-wandb.py`

**Review by frodre — APPROVED** [review]
LGTM.  Thanks for these updates.

File: external/emulation/emulation/config.py [hunk 3]
```diff
@@ -61,6 +65,12 @@ def _build_masks(self) -> Iterable[Mask]:
         for key, range in self.ranges.items():
             yield RangeMask(key, min=range.min, max=range.max)
 
+        yield partial(
```
[the remark this answers is not in the document]
Comment (nbren12) [reply external/emulation/emulation/config.py:68]: hmm...I see your point. I'm open to a rename, but can't think of a better name right now for `somefunc(state, emulator)`. "blender", "modifier"  etc seem quite vague.

File: constraints.txt [hunk 4]
```diff
@@ -49,6 +48,7 @@ commonmark==0.9.1
 conda-lock==0.4.1
 configparser==5.0.2
 crcmod==1.7
+cryptography==36.0.2
```
Comment (nbren12) [inline constraints.txt:51]: apache 2

File: constraints.txt [hunk 5]
```diff
@@ -128,6 +128,7 @@ ipython-genutils==0.2.0
 ipython==7.29.0
 ipywidgets==7.6.3
 jedi==0.18.0
+jeepney==0.8.0
```
Comment (nbren12) [inline constraints.txt:131]: MIT.

File: constraints.txt [hunk 6]
```diff
@@ -253,6 +254,7 @@ retrying==1.3.3
 rsa==4.8
 scikit-learn==1.0.2
 scipy==1.7.3
+secretstorage==3.3.1
```
Comment (nbren12) [inline constraints.txt:257]: BSD-3

File: constraints.txt [hunk 7]
```diff
@@ -299,6 +301,7 @@ tox==3.20.0
 tqdm==4.62.3
 traitlets==4.3.3
 typeguard==2.10.0
+typer==0.4.1
```
Comment (nbren12) [inline constraints.txt:304]: MIT

File: constraints.txt [hunk 8]
```diff
@@ -308,6 +311,7 @@ urllib3==1.26.8
 validators==0.18.2
 virtualenv==20.0.33
 wandb==0.12.1
+watchdog==2.1.7
```
Comment (nbren12) [inline constraints.txt:314]: Apache 2

File: constraints.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (nbren12) [reply constraints.txt:51]: Thanks. all these have compatible licenses.

**Comment by nbren12** [comment]
Thanks for the review @frodre. I hit auto-merge. The history is a bit messy, but I did launch prognostic experiments during this dev stream, so would be nice to keep them in the master branche's ancestry.

**Pull request merged**